### PR TITLE
CP-23007: Implement CD change using fd over QMP

### PR DESCRIPTION
**Pull request by fillzero**

### CODE REVIEW

**Comment by fillzero** [comment]
Given https://github.com/xapi-project/xenopsd/pull/362 was closed and the dependency has been solved, reopen this PR.

**Comment by fillzero** [comment]
@mg12 @lindig Would you please to have a look?

**Comment by lindig** [comment]
Why is Travis failing?

**Comment by fillzero** [comment]
Because this branch depends on `ocaml-qmp`,  the new added code not been released (tagged).

**Comment by lindig** [comment]
Without releasing ocaml-qmp first, why is this not going to break xenopsd? I think ocaml-qmp should be updated as part of xs-opam first and we can create a new release of xs-opam.

**Comment by fillzero** [comment]
We build a private `xs-opam` archive which fetches `ocaml-qmp`  code from the `master` branch.

File: xc/device.ml [hunk 1]
```diff
@@ -607,12 +607,41 @@ let qemu_media_change ~xs device _type params =
...
 	then begin
+		let open Qmp in
 		let cd = "ide1-cd1" in
-		let qmp_cmd =
-			if params = ""
-			then (Qmp.Command(None, Qmp.Eject (cd, Some true)))
-			else (Qmp.Command(None, Qmp.Change (cd, params, None))) in
-		qmp_write device.frontend.domid qmp_cmd
+		match params with
+		| "" ->
```
Comment (lindig) [inline xc/device.ml]: I think it is strange that we are using a string and match `""` and `_` but never use the content of that string. Doesn't this mean this is the wrong type?

**Comment by lindig** [comment]
Since this is going into a feature branch, I would not mind merging it. The code has been reviewed before.

File: xc/device.ml (hunk 1 above)
Comment (fillzero) [inline xc/device.ml]: Thanks, I use the length of string to distinguish `eject cd` and `insert cd` now.

**Comment by lindig** [comment]
Sorry, this is not better. My point is, should this not be something like a boolean or sum type with 2 components -- why this this a string in the first place if we never use it?

**Comment by fillzero** [comment]
Actually, we use this string, if it's not null, it contains the CD path, we open this path, get a file descriptor and send it to QEMU.

Otherwise If this string is empty, that means the user ejects CD.

**Pull request merged**